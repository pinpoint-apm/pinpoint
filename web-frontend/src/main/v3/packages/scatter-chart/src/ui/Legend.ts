import { SCATTER_CHART_IDENTIFIER } from '../constants/ui';
import { DataStyleMap, LegendOption } from '../types/types';

export type LegendProps = {
  legendOptions: LegendOption;
  dataStyleMap: DataStyleMap;
  width?: number;
};

export type LegendEventTypes = 'clickLegend' | 'change';

type LegendClickCallbackData = {
  checked: string[];
};

type LegendChangeChallbackData = {
  unChecked: string[];
} & LegendClickCallbackData;

type LegendEventData<T extends LegendEventTypes> = T extends 'clickLegend'
  ? LegendClickCallbackData
  : T extends 'change'
    ? LegendChangeChallbackData
    : never;

export interface LegendEventCallback<T extends LegendEventTypes> {
  (event: MouseEvent, data: LegendEventData<T>): void;
}

interface LegendEventHandlers {
  [key: string]: LegendEventCallback<LegendEventTypes>;
}

export class Legend {
  static LEGEND_CLASS = `${SCATTER_CHART_IDENTIFIER}legend`;
  static LEGEND_CONTAINER_CLASS = `${Legend.LEGEND_CLASS}_container`;
  static MARK_CLASS = `${Legend.LEGEND_CLASS}_mark`;
  static COUNT_CLASS = `${Legend.LEGEND_CLASS}_count`;
  private uniqId = new Date().getTime();
  private rootWrapper;
  private options;
  private dataStyleMap!: DataStyleMap;
  private containerElement: HTMLElement;
  private legendElements: { [key: string]: HTMLDivElement } = {};
  private eventHandlers: LegendEventHandlers = {};
  private types!: string[];
  private abortController: AbortController;

  constructor(rootWrapper: HTMLElement, { legendOptions, dataStyleMap, width }: LegendProps) {
    this.abortController = new AbortController();
    this.rootWrapper = rootWrapper;
    this.options = legendOptions;
    this.containerElement = document.createElement('div');
    this.containerElement.className = Legend.LEGEND_CONTAINER_CLASS;
    this.rootWrapper.append(this.containerElement);
    this.setSize(width);
    this.addEventListener();
    this.setDataStyleMap(dataStyleMap);
  }

  get container() {
    return this.containerElement;
  }

  public setSize(width?: number) {
    if (width) {
      this.containerElement.style.width = `${width}px`;
    } else {
      this.containerElement.style.width = `${this.rootWrapper.clientWidth}px`;
    }
  }

  public setDataStyleMap(dataStyleMap: DataStyleMap) {
    this.dataStyleMap = dataStyleMap;
    this.types = Object.keys(dataStyleMap);
    return this;
  }

  private addEventListener = () => {
    const { signal } = this.abortController;
    this.containerElement.addEventListener(
      'click',
      (event) => {
        const { target } = event;
        if (target instanceof HTMLElement) {
          const isInputNode = target.nodeName === 'INPUT';

          if (isInputNode) {
            const checkedInputElements = this.containerElement.querySelectorAll('input:checked');
            const checkedTypes = [...checkedInputElements].map((inputElement) => {
              return inputElement.getAttribute('data-name') || '';
            });

            this.eventHandlers.change?.(event, {
              checked: checkedTypes,
              unChecked: this.types.filter((type) => !checkedTypes.includes(type)),
            });
            this.eventHandlers.clickLegend?.(event, {
              checked: checkedTypes,
            });
          }
        }
      },
      { signal },
    );
  };

  public onChange(callback: LegendEventCallback<'change'>) {
    this.eventHandlers['change'] = callback;
  }

  public render() {
    const options = this.options;
    const dataTypes = this.types;

    this.containerElement.style.visibility = 'visible';
    dataTypes.forEach((type) => {
      // wrapper div
      const legendWrapper = document.createElement('div');
      legendWrapper.className = `${Legend.LEGEND_CLASS} ${type}`;

      // mark
      const markElement = document.createElement('span');
      markElement.style.background = this.dataStyleMap[type].legend;
      // const formattedLabel = options?.formatLabel?.(type) || type;
      markElement.className = Legend.MARK_CLASS;

      // count span
      const countElement = document.createElement('span');
      countElement.className = Legend.COUNT_CLASS;
      countElement.innerHTML = '0';

      // label
      const labelElement = document.createElement('label');
      const formattedLabel = options?.formatLabel?.(type) || type;
      labelElement.htmlFor = `${Legend.LEGEND_CLASS}_${type}_input_${this.uniqId}`;
      labelElement.append(`${formattedLabel}`, countElement);

      // input
      const inputElement = document.createElement('input');
      inputElement.id = `${Legend.LEGEND_CLASS}_${type}_input_${this.uniqId}`;
      inputElement.type = 'checkbox';
      inputElement.dataset.name = type;
      inputElement.checked = true;

      this.legendElements[type] = legendWrapper;

      legendWrapper.append(markElement, labelElement, inputElement);
      this.containerElement.append(legendWrapper);
    });
    return this;
  }

  public unmount() {
    const containerElement = this.containerElement;
    while (containerElement.firstChild) {
      containerElement.removeChild(containerElement.firstChild);
    }
    return this;
  }

  public setLegendCount(type: string, value: number) {
    const legendElement = this.legendElements[type];
    const countElement = legendElement.getElementsByClassName(Legend.COUNT_CLASS)[0];

    countElement.innerHTML = `${this.options.formatValue?.(value)}`;
  }

  public on(eventType: LegendEventTypes, callback: LegendEventCallback<'clickLegend'>) {
    this.eventHandlers[eventType] = callback;
  }

  public off(eventType: LegendEventTypes) {
    delete this.eventHandlers[eventType];
  }

  public destroy() {
    this.abortController.abort();
    this.unmount();
    this.containerElement.remove();
    const keys = Object.keys(this.eventHandlers) as LegendEventTypes[];
    keys.forEach((key: LegendEventTypes) => {
      this.off(key);
    });
  }
}
