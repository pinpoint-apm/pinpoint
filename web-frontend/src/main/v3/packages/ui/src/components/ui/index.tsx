export * from './accordion';
export * from './alert';
export * from './alert-dialog';
export * from './badge';
export * from './button';
export * from './button-group';
export * from './card';
export * from './chart';
export * from './checkbox';
export * from './collapsible';
export * from './command';
export * from './dialog';
export * from './dropdown-menu';
export * from './form';
export * from './hover-card';
export * from './input';
export * from './label';
export * from './popover';
export * from './radio-group';
export * from './resizable';
export * from './scroll-area';
export * from './select';
export * from './separator';
export * from './sheet';
export * from './table';
export * from './textarea';
export * from './toast';
export * from './toaster';
export * from './tooltip';
export * from './tabs';
