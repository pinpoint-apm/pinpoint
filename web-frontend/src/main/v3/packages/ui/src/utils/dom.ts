export const hasParentWithId = (element: HTMLElement | null, parentId: string): boolean => {
  if (!element) {
    return false;
  }
  if (element.id === parentId) {
    return true;
  }
  return hasParentWithId(element.parentElement, parentId);
};
