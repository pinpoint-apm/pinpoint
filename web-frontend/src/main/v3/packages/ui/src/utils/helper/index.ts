export * from './application';
export * from './charts';
export * from './filteredMap/index';
export * from './route';
export * from './routeV2';
export * from './scatter';
export * from './serverMap';
export * from './transaction';
