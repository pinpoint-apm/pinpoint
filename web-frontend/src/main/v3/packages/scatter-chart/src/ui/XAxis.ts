import { DeepNonNullable, TickOption } from '../types/types';
import { drawLine, drawText } from '../utils/draw';
import { Axis, AxisProps } from './Axis';

export class XAxis extends Axis {
  constructor({ ...props }: AxisProps) {
    super(props);
    this.priority = -1;
  }

  public render() {
    this.clear();
    const { min, max, tick, innerPadding, strokeColor } = this;
    const {
      format,
      count,
      color,
      width: tickWidth,
      strokeColor: tickStrokeColor,
    } = tick as DeepNonNullable<TickOption>;
    const padding = this.padding;
    const width = this.canvas.width / this.dpr;
    const height = this.canvas.height / this.dpr;
    const startX = padding.left + innerPadding;
    const endX = width - padding.right - innerPadding;
    const endY = height - padding.bottom;
    const wGap = (endX - startX) / (count - 1);
    const xTickGap = (max - min) / (count - 1);
    this.setStyle();

    [...Array(count)].forEach((_, i) => {
      const x = wGap * i + startX;
      const label = `${format(xTickGap * i + min)}`;
      const lines = `${label}`.split('\n');
      const textHeight = this.getTextHeight(label);

      lines.reverse().forEach((line, i) => {
        drawText(
          this.context,
          `${line}`,
          x,
          height - padding.bottom + tick!.width! + textHeight + tick!.padding!.top! - i * this.getTextHeight(line),
          { textAlign: 'center', textBaseline: 'bottom', color },
        );
      });
      drawLine(this.context, x, endY, x, endY + tickWidth, { color: tickStrokeColor });
    });
    drawLine(this.context, startX - innerPadding, endY, endX + innerPadding, endY, { color: strokeColor });
  }
}
