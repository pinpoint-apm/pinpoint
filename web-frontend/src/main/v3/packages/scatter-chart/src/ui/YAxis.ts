import { DeepNonNullable, TickOption } from '../types/types';
import { drawLine, drawRect, drawText } from '../utils/draw';
import { Axis, AxisProps } from './Axis';

export interface YAxisProps extends AxisProps {
  backgroundColor?: string;
}

export class YAxis extends Axis {
  private backgroundColor;

  constructor({ backgroundColor, ...props }: YAxisProps) {
    super(props);
    this.priority = -2;
    this.isFixed = true;
    this.backgroundColor = backgroundColor;
  }

  public render() {
    this.clear();
    const { min, max, tick, innerPadding, backgroundColor, strokeColor } = this;
    const {
      format,
      count,
      color,
      width: tickWidth,
      strokeColor: tickStrokeColor,
    } = tick as DeepNonNullable<TickOption>;
    const padding = this.padding;
    const width = this.canvas.width / this.dpr;
    const height = this.canvas.height / this.dpr;

    const startX = padding.left;
    const startY = padding.top + innerPadding;
    const endY = height - padding.bottom - innerPadding;
    const hGap = (endY - startY) / (count - 1);
    const yTickGap = (max - min) / (count - 1);
    this.setStyle();

    drawRect(this.context, 0, 0, padding.left, endY + innerPadding + tickWidth, { color: backgroundColor });
    drawRect(this.context, width - padding.right, 0, width, endY + innerPadding + tickWidth, {
      color: backgroundColor,
    });
    drawRect(this.context, 0, 0, width, padding.top, { color: backgroundColor });
    [...Array(count)].forEach((_, i) => {
      const y = hGap * i + startY;
      const label = `${format(yTickGap * (count - 1 - i) + min)}`;
      drawText(
        this.context,
        `${label}`,
        startX - tick!.padding!.right! - tickWidth,
        y + this.getTextHeight(label) / 4,
        { textAlign: 'end', color },
      );
      drawLine(this.context, startX - tickWidth, y, startX, y, { color: tickStrokeColor });
    });
    drawLine(this.context, startX, startY - innerPadding, startX, endY + innerPadding, { color: strokeColor });
  }
}
