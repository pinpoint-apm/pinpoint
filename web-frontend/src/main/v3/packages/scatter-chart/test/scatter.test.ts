import 'jest-canvas-mock';
import { fireEvent, waitFor } from '@testing-library/dom';
import data1 from './mock/data1.json';
import data2 from './mock/data2.json';
import { initOption, ScatterChartTestHelper, simulateDrag } from './helper';
import { Legend } from '../src/ui/Legend';
import { Viewport } from '../src/ui/Viewport';

describe('Test for Scatter', () => {
  let SC: ScatterChartTestHelper;
  const wrapper = document.createElement('div');

  describe('Test for Scatter Method', () => {
    it('should check data rendered by `render` method', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);

      // when
      SC.render(data1.data);

      // then
      const events = SC.viewport.context.__getEvents();
      expect(events).toMatchSnapshot();
    });

    it('should check data rendered by `render` method with append flag is true', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);

      // when
      SC.render(data1.data, { append: true });
      SC.render(data2.data, { append: true });

      // then
      const events = SC.viewport.context.__getEvents();
      expect(events).toMatchSnapshot();
    });

    it('should render all data then greater than y.min when drawOutOfRange flag is true', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);
      const dataCount = data1.data.filter((d) => d.y > SC.getYAxis().min && d.type === 'success').length;

      // when
      SC.render(data1.data, { drawOutOfRange: true });
      const successCount = wrapper.querySelector('.success .__scatter_chart__legend_count')?.innerHTML;

      // then
      waitFor(() => {
        expect(successCount).toBe(`${dataCount}`);
      });
    });

    it('should occur `click` callback function registerd by `on` method when clicking chart area', () => {
      // given
      const onClick = jest.fn();
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.on('click', onClick);

      // when
      fireEvent.click(SC.getGuide()!.canvas);

      // then
      expect(onClick).toHaveBeenCalled();
    });

    it.skip('should occur `dragEnd` callback function registerd by `on` method when drag chart area', () => {
      // given
      const onDragEnd = jest.fn();
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.on('dragEnd', onDragEnd);

      // when
      simulateDrag(SC.getGuide()!.canvas, 0, 0, 100, 100);

      // then
      expect(onDragEnd).toHaveBeenCalled();
    });

    it('should occur `resize` callback function registerd by `on` method when resize method called', () => {
      // given
      const onResize = jest.fn();
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.on('resize', onResize);

      // when
      SC.resize(100, 100);

      // then
      expect(onResize).toHaveBeenCalled();
    });

    it('should occur `clickLegend` callback function registerd by `on` method when clicking legend area', () => {
      // given
      const onClick = jest.fn();
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.on('clickLegend', onClick);

      // when
      const inputElement = SC.getLegend()!.container.getElementsByTagName('input')[0];
      fireEvent.click(inputElement);

      // then
      expect(onClick).toHaveBeenCalled();
      expect(onClick.mock.calls[0][1]).toMatchObject({ checked: ['fail'] });
    });

    it('should not occur callback function when remove callback event by `off` method', () => {
      // given
      const onClick = jest.fn();
      const onClickLegend = jest.fn();
      const onDragEnd = jest.fn();
      const onResize = jest.fn();
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.on('click', onClick);
      SC.on('dragEnd', onDragEnd);
      SC.on('clickLegend', onClickLegend);
      SC.on('resize', onResize);
      SC.off('click');
      SC.off('dragEnd');
      SC.off('clickLegend');
      SC.off('resize');

      // when
      SC.resize(100, 100);
      fireEvent.click(SC.getGuide()!.canvas);

      // then
      expect(onClick).not.toHaveBeenCalled();
      expect(onClickLegend).not.toHaveBeenCalled();
      expect(onDragEnd).not.toHaveBeenCalled();
      expect(onResize).not.toHaveBeenCalled();
    });

    it('should be resized by `resize` method', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);

      // when
      SC.resize(100, 100);

      // then
      expect(SC.viewport.styleWidth).toBe(100);
      expect(SC.viewport.styleHeight).toBe(100);
    });

    it('data should be cleared by `clear` method', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);

      // when
      SC.clear();

      // then
      expect(SC.getData().length).toBe(0);
    });

    it('all legends should be 0 by `clear` method', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.render(data1.data);

      // when
      SC.clear();

      // then
      const successCount = wrapper.querySelector('.success .__scatter_chart__legend_count')?.innerHTML;
      const failCount = wrapper.querySelector('.fail .__scatter_chart__legend_count')?.innerHTML;
      waitFor(() => {
        expect(successCount).toBe('0');
        expect(failCount).toBe('0');
      });
    });

    it('should remove rendered elements(view, legend) from the screen when destroy method is called.', () => {
      // given
      SC = new ScatterChartTestHelper(wrapper, initOption);
      SC.render(data1.data);

      // when
      SC.destroy();

      // then
      const viewContainer = wrapper.querySelector(Viewport.VIEW_CONTAINER_CLASS);
      const legendContainer = wrapper.querySelector(Legend.LEGEND_CONTAINER_CLASS);

      waitFor(() => {
        expect(viewContainer?.childNodes.length).toBe('0');
        expect(legendContainer?.childNodes.length).toBe('0');
      });
    });
  });
});
