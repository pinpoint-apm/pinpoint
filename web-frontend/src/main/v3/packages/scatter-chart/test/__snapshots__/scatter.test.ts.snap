// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Scatter Test for Scatter Method should check data rendered by \`render\` method 1`] = `
Array [
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`Test for Scatter Test for Scatter Method should check data rendered by \`render\` method with append flag is true 1`] = `
Array [
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 320,
      "width": 518,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": -0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 320,
      "dWidth": 518,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="320"
        style="width: 518px; height: 320px;"
        width="518"
      />,
      "sHeight": 320,
      "sWidth": 518,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
